src/__tests__/analytics-pre-init.integration.test.ts

```diff
@@ -0,0 +1,341 @@
+      expect(consoleErrorSpy).toBeCalledWith('identity rejection')
+
+      expect(onSpy).toBeCalledTimes(1)
+
+      expect(onTrackCb).toBeCalledTimes(2) // gets called once for each track event
+      expect(onTrackCb).toBeCalledWith('foo', {}, undefined)
+      expect(onTrackCb).toBeCalledWith('bar', {}, undefined)
+    })
+  })
+
+  describe('Emitter methods', () => {
+    test('If, before initialization, .on("track") is called, the .on method should be called after analytics load', async () => {
+      const ajsBuffered = AnalyticsBrowser.load({ writeKey })
+      const args = ['track', jest.fn()] as const
+      ajsBuffered.on(...args)
+      expect(onSpy).not.toHaveBeenCalledWith(...args)
+
+      await ajsBuffered
+      expect(onSpy).toBeCalledWith(...args)
+      expect(onSpy).toHaveBeenCalledTimes(1)
+    })
+
+    test('If, before initialization .on("track") is called and then .track is called, the callback method should be called after analytics loads', async () => {
+      const onFnCb = jest.fn()
+      const analytics = AnalyticsBrowser.load({ writeKey })
+      analytics.on('track', onFnCb)
+      const trackCtxPromise = analytics.track('foo', { name: 123 })
+
+      expect(onFnCb).not.toHaveBeenCalled()
+
+      await Promise.all([analytics, trackCtxPromise])
+
+      expect(onSpy).toBeCalledWith('track', onFnCb)
+      expect(onSpy).toHaveBeenCalledTimes(1)
+
+      expect(onFnCb).toHaveBeenCalledWith('foo', { name: 123 }, undefined)
+      expect(onFnCb).toHaveBeenCalledTimes(1)
+    })
+
+    test('If, before initialization, .ready is called, the callback method should be called after analytics loads', async () => {
+      const onReadyCb = jest.fn()
+      const analytics = AnalyticsBrowser.load({ writeKey })
+      const onReadyPromise = analytics.ready(onReadyCb)
+      expect(onReadyCb).not.toHaveBeenCalled()
+      await onReadyPromise
+      expect(readySpy).toHaveBeenCalledTimes(1)
+      expect(onReadyCb).toHaveBeenCalledTimes(1)
+      expect(readySpy).toHaveBeenCalledWith(expect.any(Function))
+    })
+
+    test('Should work with "on" events if a track event is called after load is complete', async () => {
+      const onTrackCb = jest.fn()
+      const ajsBuffered = AnalyticsBrowser.load({ writeKey })
+      ajsBuffered.on('track', onTrackCb)
+      await ajsBuffered
+      await ajsBuffered.track('foo', { name: 123 })
+
+      expect(onTrackCb).toHaveBeenCalledTimes(1)
+      expect(onTrackCb).toHaveBeenCalledWith('foo', { name: 123 }, undefined)
+    })
+    test('"on, off, once" should return ajsBuffered', () => {
+      const analytics = AnalyticsBrowser.load({ writeKey })
+      expect(
+        [
+          analytics.on('track', jest.fn),
+          analytics.off('track', jest.fn),
+          analytics.once('track', jest.fn),
+        ].map((el) => el instanceof AnalyticsBuffered)
+      ).toEqual([true, true, true])
+    })
+
+    test('"emitted" events should be chainable', async () => {
```

Comment:
Can we have a test that also ensures that `this` passed to the event listener callbacks is the `analytics` instance? It looks like it is from my testing, I just want to make sure we don't accidentally break that in the future.